### PR TITLE
superfluous response.WriteHeader warning for Gzip

### PR DESCRIPTION
**Pull request by bartoval**
Roll back to the previous implementation of gzip to avoid the warning message triggered by the Gorilla Mux library.

### CODE REVIEW

File: cmd/flow-collector/main.go [hunk 1]
```diff
@@ -383,18 +408,15 @@ func main() {
...
 		if exists {
 			userResponse := handler(r)
 			if response, err := json.Marshal(userResponse); err == nil {
-				statusCode = http.StatusOK
 				fmt.Fprintf(w, "%s", response)
 			}
 		}
 
 		w.Header().Set("Content-Type", "application/json")
-		w.WriteHeader(statusCode)
```
Comment (bartoval) [inline cmd/flow-collector/main.go:397]: I also received the same warning also for this statement

**Comment by c-kruse** [comment]
@bartoval curious about what the warnings are you are seeing are!

**Comment by bartoval** [comment]
> @bartoval curious about what the warnings are you are seeing are!

Flow collector logs
p

**Review by c-kruse — COMMENTED** [review]
@bartoval  I had a hard time reproducing this, but I think you found the root of the issue on line 397 (calling http.ResponseWriter.WriteHeader after having already written the response on line 392.) I would be terribly surprised if the gorilla gzip handler was doing something wrong here - it appears to me that it just gets the blame because it wraps the ResponseWriter.

**Comment by bartoval** [comment]
> @bartoval I had a hard time reproducing this, but I think you found the root of the issue on line 397 (calling http.ResponseWriter.WriteHeader after having already written the response on line 392.) I would be terribly surprised if the gorilla gzip handler was doing something wrong here - it appears to me that it just gets the blame because it wraps the ResponseWriter.

👍🏼  You are right. 
The only API affected by this warning is /user. It was like using an axe for a surgical operation.

At this point, however, the doubt is: what is the difference between this API and the others? All of them explicitly call 'responseWrite' in the controller.

For example, this one: 
`var siteApi = api1.PathPrefix("/sites").Subrouter()
	siteApi.StrictSlash(true)
	**siteApi.HandleFunc("/", authenticated(http.HandlerFunc(c.siteHandler))).Name("list")**`

that call this function:

`func (c *Controller) siteHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	c.FlowCollector.Request <- flow.ApiRequest{RecordType: flow.Site, Request: r}
	response := <-c.FlowCollector.Response
	**w.WriteHeader(response.Status)**
	if response.Body != nil {
		fmt.Fprintf(w, "%s", *response.Body)
	}
}`

**Comment by c-kruse** [comment]
>At this point, however, the doubt is: what is the difference between this API and the others? All of them explicitly call 'responseWrite'.

The other API handlers write headers first, then they write the body - it's the order that was the problem with /users. When we call fmt.Fprintf(w, ) we are really just indirectly calling http.ResponseWriter.Write. The go docs for that function try to describe it. Basically go expects you to have set any status and additional headers before you begin to write the body, otherwise it will make some guesses quickly to start sending the response headers and body.
```
~ » go doc net/http.ResponseWriter.Write
package http // import "net/http"

type ResponseWriter interface {

        // Write writes the data to the connection as part of an HTTP reply.
        //
        // If WriteHeader has not yet been called, Write calls
        // WriteHeader(http.StatusOK) before writing the data. If the Header
        // does not contain a Content-Type line, Write adds a Content-Type set
        // to the result of passing the initial 512 bytes of written data to
        // DetectContentType. Additionally...
```

So preserving the original intended behavior, I'd probably do something more like this:

if !exists {
  w.WriteHeader(204)
  return
}
response, err = unmarshal resposne
if err != nil { // maybe log and 500? we shouldn't really ever get there
  return
}
w.Header().Set("Content-Type", "application/json")
w.Write(response)

**Comment by bartoval** [comment]
> > At this point, however, the doubt is: what is the difference between this API and the others? All of them explicitly call 'responseWrite'.
> 
> The other API handlers write headers first, then they write the body - it's the order that was the problem with /users. When we call fmt.Fprintf(w, ) we are really just indirectly calling http.ResponseWriter.Write. The go docs for that function try to describe it. Basically go expects you to have set any status and additional headers before you begin to write the body, otherwise it will make some guesses quickly to start sending the response headers and body.
> 
> ```
> ~ » go doc net/http.ResponseWriter.Write
> package http // import "net/http"
> 
> type ResponseWriter interface {
> 
>         // Write writes the data to the connection as part of an HTTP reply.
>         //
>         // If WriteHeader has not yet been called, Write calls
>         // WriteHeader(http.StatusOK) before writing the data. If the Header
>         // does not contain a Content-Type line, Write adds a Content-Type set
>         // to the result of passing the initial 512 bytes of written data to
>         // DetectContentType. Additionally...
>        Write([]byte) (int, error)
> }
> ```
> 
> So preserving the original intended behavior, I'd probably do something more like this:
> 
> if !exists { w.WriteHeader(204) return } response, err = unmarshal resposne if err != nil { // maybe log and 500? we shouldn't really ever get there return } w.Header().Set("Content-Type", "application/json") w.Write(response)

Perfect. Updated the PR.
Thanks for the clarification 👍🏼

**Review by c-kruse — APPROVED** [review]
Looks good to me! Thank you.

**Pull request merged**